db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,129 @@
+    }
+    return std::make_pair(directory_name.substr(0, pos), table_id(utils::UUID(directory_name.substr(pos + 1))));
+}
+
+void backup_task_impl::run_in_thread(sstring data_dir) {
+    std::exception_ptr ex;
+    //
+    // The keyspace data directories and their snapshots are arranged as follows:
+    //
+    //  <data dir>
+    //  |- <keyspace name1>
+    //  |  |- <column family name1>
+    //  |     |- snapshots
+    //  |        |- <snapshot name1>
+    //  |          |- <snapshot file1>
+    //  |          |- <snapshot file2>
+    //  |          |- ...
+    //  |        |- <snapshot name2>
+    //  |        |- ...
+    //  |  |- <column family name2>
+    //  |  |- ...
+    //  |- <keyspace name2>
+    //  |- ...
+    //
+    auto ks_dir = fs::path(data_dir) / _ks;
+    auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+    auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+    snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+    while (auto table_ent = ks_dir_lister.get().get()) {
+        auto snapshot_dir = ks_dir/ table_ent->name / sstables::snapshots_dir / _snapshot_name;
+        auto has_snapshot = file_exists(snapshot_dir.native()).get();
+
+        snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
+        if (!has_snapshot) {
+            continue;
+        }
+
+        auto cf_name_and_uuid = extract_cf_name_and_uuid(table_ent->name);
+        auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+        auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
+
+        while (auto component_ent = snapshot_dir_lister.get().get()) {
+            auto gh = _uploads.hold();
+            auto component_name = snapshot_dir / component_ent->name;
+            auto destination = fmt::format("/{}/{}/{}/{}", _bucket, cf_name_and_uuid.first, _snapshot_name, component_ent->name);
+            snap_log.trace("Upload {} to {}", component_name.native(), destination);
+            std::ignore = _client->upload_file(component_name, destination).handle_exception([this, comp = component_name] (auto ex) {
+                snap_log.error("Error uploading {}: {}", comp.native(), ex);
+                _ex = std::move(ex);
+            }).finally([gh = std::move(gh)] {});
+            thread::maybe_yield();
+            if (_ex) {
+                break;
+            }
+        }
+    }
+}
+
+future<> backup_task_impl::run() {
+    co_await _snap_ctl.run_snapshot_list_operation(coroutine::lambda([this] -> future<> {
+        std::vector<sstring> data_dirs = _snap_ctl._db.local().get_config().data_file_directories();
+        co_await coroutine::parallel_for_each(data_dirs, [this] (auto& d) {
+            return async([this, d] {
+                try {
+                    run_in_thread(d);
```

Comment:
why isn't the thread created there (in `run_in_thread`)?

Reply:
To save up two indentation levels -- one for async lambda and another one for surrounding try-catch

Reply:
It'd be better to encapsulate it there so it will be clear that it's this function choice to launch a thread.

Reply:
as for the try catch, why isn't it possible in the coroutine lambda?

Reply:
The `..._in_thread()` in a method name is pretty standard already. The most prominent example is consume_in_thread(), lot's of such calls over there

Reply:
> as for the try catch, why isn't it possible in the coroutine lambda?

I don't follow. It is in lambda already, or is your proposal in

```c++
    co_await parallel_for_each([] {
        try {
            return async([] {
                run_in_thread();
            });
        } catch (...) {
        }
    });
```

?

Reply:
> The `..._in_thread()` in a method name is pretty standard already. The most prominent example is consume_in_thread(), lot's of such calls over there

True, but the reason for that is that we started with a thread up in the stack and relied on that when writing the helper and lower-level functions.  This is addictive and in general we want to use coroutines to try to avoid that.

In this case we start from scratch and the only reason to have this thread is to simplify what is now `run_in_thread`.  Therefore I think it should be started there as a local choice of this function.

> > as for the try catch, why isn't it possible in the coroutine lambda?
>
> I don't follow. It is in lambda already, or is your proposal in

Yes, it can remain in the lambda like:
```
co_await coroutine::parallel_for_each([] {
    try {
        co_await run();
    } catch (...) {
        _ex = ...
    }
});
```

My comment was about saving the indentation levels :)